### PR TITLE
storage/filesystem: call initialization explicitly, fixes #408

### LINKED ISSUE #408: PlainOpen on non-git directory creates git directory structure
Calling `PlainOpen` on a directory path that is not a git repository has a side effect of creating the git directory structure within the directory:
```
├── objects
│   ├── info
│   └── pack
└── refs
    ├── heads
    └── tags
```

I would expect the following test to pass, but it fails as follows:
```go
func (s *RepositorySuite) TestPlainOpenEmptyDir(c *C) {
	dir, err := ioutil.TempDir("", "plain-open-empty-dir")
	c.Assert(err, IsNil)
	defer os.RemoveAll(dir)

	r, err := PlainOpen(dir)
	c.Assert(err, Equals, ErrRepositoryNotExists)
	c.Assert(r, IsNil)

	entries, err := ioutil.ReadDir(dir)
	c.Assert(err, IsNil)
	c.Assert(len(entries), Equals, 0)
}
```

However, it currently fails:
```
FAIL: repository_test.go:368: RepositorySuite.TestPlainOpenEmptyDir

repository_test.go:379:
    c.Assert(len(entries), Equals, 0)
... obtained int = 2
... expected int = 0

OOPS: 139 passed, 1 FAILED
--- FAIL: Test (33.04s)
```

### PR DESCRIPTION
**Pull request by smola**
filesystem.Storage was initializing the gitdir (creating objects
and refs) on NewStorage. But this should be done only on init and
clone operations, not on open.

Now there is a new interface storer.Initializer that storers can
implement if they need any initialization step before init or clone.
filesystem.Storage is one of such implementations.

git.Init and git.Clone now call to the storer Init() method if it
does implement it. Otherwise, it just ignores initialization.

### CODE REVIEW

**Comment by smola** [comment]
Fixes https://github.com/src-d/go-git/issues/408

File: storage/filesystem/storage_test.go [hunk 1]
```diff
@@ -8,18 +8,21 @@ import (
...
 	"gopkg.in/src-d/go-billy.v2/memfs"
 	"gopkg.in/src-d/go-billy.v2/osfs"
+	"io/ioutil"
 )
 
 func Test(t *testing.T) { TestingT(t) }
 
 type StorageSuite struct {
 	test.BaseStorageSuite
+	Dir string
```
Comment (alcortesm) [inline storage/filesystem/storage_test.go]: Let's make this attribute private rather than public.

File: storage/filesystem/storage_test.go [hunk 2]
```diff
@@ -42,3 +45,9 @@ func (s *StorageSuite) TestFilesystem(c *C) {
 
 	c.Assert(storage.Filesystem(), Equals, fs)
 }
+
+func (s *StorageSuite) TestStructureShouldBeCreatedLazily(c *C) {
```
Comment (alcortesm) [inline storage/filesystem/storage_test.go]: The test name is misleading, what Structure?, and we are also not testing that contents are created lazily, just that the dir is created empty.

I suggest a new name for the test: `TestNewStorageShouldNotAddAnyContentsToDir`.

**Comment by alcortesm** [comment]
Of course storage_test.go:31: StorageSuite.TestNewStorage is failing now, that was the whole point of the patch :).  By the way, the name of that test was not very informative :).

**Comment by ajnavarro** [comment]
@alcortesm What I meant is, if the behavior has changed, this test should not test if the directories are created, so this test should pass too (removing the Stats line, the entire test because now does not make sense, or whatever)

**Comment by alcortesm** [comment]
@ajnavarro yes, I understood you and I think you are right.  My comment was just to point out that the name of the test and the things it tested were not aligned.

I also agree with you that that whole test makes no sense.

File: storage/filesystem/storage_test.go (hunk 1 above)
Comment (smola) [inline storage/filesystem/storage_test.go]: :+1:

File: storage/filesystem/storage_test.go (hunk 2 above)
Comment (smola) [inline storage/filesystem/storage_test.go]: Renamed.

**Comment by smola** [comment]
@alcortesm @ajnavarro Right. I removed `TestNewStorage`, we want exactly the oposite behavior.

**Comment by alcortesm** [comment]
@mcuadros 

What do you mean by "the structure" and why it should be there?

Can you write a test to illustrate exactly what **has** to be there and what is not really needed until we use it for something?

**Comment by smola** [comment]
@mcuadros and me talked about this offline and agreed that this should be done explicitly. That is: a storer needs a way to get called when initialization is being done.

The end result is essentially the same: as you see, the test has not changed.
The difference, however, is that now `git.Init` (and `git.Clone` indirectly) explicitly calls `Init()` on the storer if it implements it.

**Comment by alcortesm** [comment]
I see, thanks!

File: storage/filesystem/storage_test.go [hunk 3]
```diff
@@ -14,31 +15,29 @@ func Test(t *testing.T) { TestingT(t) }
...
 }
 
-func (s *StorageSuite) TestNewStorage(c *C) {
+func (s *StorageSuite) TestFilesystem(c *C) {
 	fs := memfs.New()
 	storage, err := NewStorage(fs)
 	c.Assert(err, IsNil)
-	c.Assert(storage, NotNil)
 
-	_, err = fs.Stat("refs/tags")
```
Comment (mcuadros) [inline storage/filesystem/storage_test.go]: why we delete this test?

Comment (smola) [inline storage/filesystem/storage_test.go]: Its only purpose was checking that refs/tags was created before any action. This is no longer the case, so it's enough if we check the other cases.
